packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -476,223 +489,319 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
+    return _RenderTheatre(
+      skipCount: skipCount,
+      textDirection: Directionality.of(context),
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderTheatre renderObject) {
+    renderObject
+      ..skipCount = skipCount
+      ..textDirection = Directionality.of(context);
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(IntProperty('skipCount', skipCount));
+  }
 }
 
-class _TheatreElement extends RenderObjectElement {
-  _TheatreElement(_Theatre widget)
-    : assert(!debugChildrenHaveDuplicateKeys(widget, widget.offstage)),
-      super(widget);
+class _TheatreElement extends MultiChildRenderObjectElement {
+  _TheatreElement(_Theatre widget) : super(widget);
 
   @override
   _Theatre get widget => super.widget as _Theatre;
 
   @override
   _RenderTheatre get renderObject => super.renderObject as _RenderTheatre;
 
-  Element _onstage;
-  static final Object _onstageSlot = Object();
+  @override
+  void debugVisitOnstageChildren(ElementVisitor visitor) {
+    assert(children.length >= widget.skipCount);
+    children.skip(widget.skipCount).forEach(visitor);
+  }
+}
+
+class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox, StackParentData> {
+  _RenderTheatre({
+    List<RenderBox> children,
+    @required TextDirection textDirection,
+    int skipCount = 0,
+  }) : assert(skipCount != null),
+       assert(skipCount >= 0),
+       assert(textDirection != null),
+       _textDirection = textDirection,
+       _skipCount = skipCount {
+    addAll(children);
+  }
 
-  List<Element> _offstage;
-  final Set<Element> _forgottenOffstageChildren = HashSet<Element>();
+  bool _hasVisualOverflow = false;
 
   @override
-  void insertChildRenderObject(RenderBox child, dynamic slot) {
-    assert(renderObject.debugValidateChild(child));
-    if (slot == _onstageSlot) {
-      assert(child is RenderStack);
-      renderObject.child = child as RenderStack;
-    } else {
-      assert(slot == null || slot is Element);
-      renderObject.insert(child, after: slot?.renderObject as RenderBox);
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! StackParentData)
+      child.parentData = StackParentData();
+  }
+
+  Alignment _resolvedAlignment;
+
+  void _resolve() {
+    if (_resolvedAlignment != null)
+      return;
+    _resolvedAlignment = AlignmentDirectional.topStart.resolve(textDirection);
+  }
+
+  void _markNeedResolution() {
+    _resolvedAlignment = null;
+    markNeedsLayout();
+  }
+
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection == value)
+      return;
+    _textDirection = value;
+    _markNeedResolution();
+  }
+
+  int get skipCount => _skipCount;
+  int _skipCount;
+  set skipCount(int value) {
+    assert(value != null);
```

Comment:
assert that the children we're skipping are offstage?

Reply:
Same answer as above: The fact that we are skipping them makes them offstage. Not the other way around.